File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        result = []

        def dfs(curr, greatest):
            if not curr:
                return None
            
            if curr.val >= greatest:
                result.append(curr.val)
                greatest = curr.val
            dfs(curr.left, greatest)
            dfs(curr.right, greatest)
        dfs(root, root.val)

        return len(result)

File: Data Structures & Algorithms/delete-node-in-a-bst/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        if (root == null){
            return root;
        }
        if (root.val == key){
            if (root.left == null) {
                return root.right;
            } else if (root.right == null){
                return root.left;
            }

            TreeNode curr = root.right;
            while(curr.left != null){
                curr = curr.left;
            }

            curr.left = root.left;
            TreeNode result = root.right;
            root = null;
            return result;

        } else if (root.val < key){
            root.right = deleteNode(root.right, key);
        } else {
            root.left = deleteNode(root.left, key);
        }
        return root;
    }
}